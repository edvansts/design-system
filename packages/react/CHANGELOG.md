# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.4.1](https://github.com/edvansts/design-system/compare/v0.4.0...v0.4.1) (2021-12-09)


### Bug Fixes

* :heavy_plus_sign: add @matoss/scss in package json react ([01dc5d4](https://github.com/edvansts/design-system/commit/01dc5d4fa98628394375ac69542250afae271662))





# [0.4.0](https://github.com/edvansts/design-system/compare/v0.3.0...v0.4.0) (2021-12-08)


### Features

* add commitizen ([db9bd6b](https://github.com/edvansts/design-system/commit/db9bd6b9bd20518cdd609943faffb6bbaa5b7e7e))





# [0.3.0](https://github.com/edvansts/design-system/compare/v0.2.0...v0.3.0) (2021-12-08)


### Bug Fixes

* :bug: react version'' ([00ae03b](https://github.com/edvansts/design-system/commit/00ae03be10e17622b10fe69581570af60103be21))


### Features

* :poop: test ([7c12185](https://github.com/edvansts/design-system/commit/7c1218583a5d9565dda7d02d99f041fac30a3535))





# [0.2.0](https://github.com/edvansts/design-system/compare/v0.1.1...v0.2.0) (2021-12-07)


### Bug Fixes

* :bug: adjusting dependencies ([8983b15](https://github.com/edvansts/design-system/commit/8983b158a2986b0e8dfdb56fc97b01dbea1f3af1))
* files for package.json files ([0f0f790](https://github.com/edvansts/design-system/commit/0f0f790b3847d2f99095b560da5b090772eb3ddb))


### Features

* :white_check_mark: private repo ([4a96c97](https://github.com/edvansts/design-system/commit/4a96c978cbb6fa972e0637eca84c2946504b80af))
