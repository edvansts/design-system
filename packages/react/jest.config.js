module.exports = {
  roots: ["<rootDir>/src"],
  testRegex: "(/.*\\.test)\\.(ts|tsx)$",
  setupFilesAfterEnv: ["<rootDir>/src/setupTests.ts"],
  verbose: true
};
