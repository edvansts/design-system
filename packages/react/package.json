{
  "name": "@matoss/react",
  "version": "0.4.1",
  "main": "lib/index.js",
  "license": "MIT",
  "files": [
    "lib"
  ],
  "devDependencies": {
    "@babel/preset-env": "^7.10.4",
    "@babel/preset-react": "^7.10.4",
    "@babel/preset-typescript": "^7.10.4",
    "@matoss/foundation": "^0.4.1",
    "@storybook/addon-a11y": "^5.3.19",
    "@storybook/addon-knobs": "^5.3.19",
    "@storybook/addon-storysource": "^5.3.19",
    "@storybook/preset-typescript": "^3.0.0",
    "@storybook/react": "^5.3.19",
    "@testing-library/jest-dom": "^5.11.1",
    "@testing-library/react": "^10.4.7",
    "@testing-library/user-event": "^13.5.0",
    "@types/jest": "^26.0.7",
    "@types/react": "^17.0.37",
    "@types/react-dom": "^17.0.11",
    "babel-loader": "^8.1.0",
    "chromatic": "^5.0.0",
    "jest": "^26.1.0",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "rollup": "^2.60.2",
    "rollup-plugin-typescript2": "^0.31.1"
  },
  "scripts": {
    "build": "rollup -c",
    "dev": "yarn build --watch",
    "test": "jest --verbose",
    "test:watch": "yarn test --watch",
    "storybook:start": "start-storybook",
    "storybook:build": "build-storybook",
    "chromatic": "chromatic"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "12eeba8c70b7f7175e882cf91fc46fc0218a0f1d"
}
