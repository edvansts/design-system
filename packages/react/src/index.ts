import Button, { ButtonProps } from "./atoms/Button";
import Color, { ColorProps } from "./atoms/Color";
import Spacing, { SpacingTypes } from "./foundation/Spacing";
import Image, { ImageProps } from "./atoms/Image";

export { Button, Color, Spacing, Image };

export type { ButtonProps, ColorProps, SpacingTypes, ImageProps };
