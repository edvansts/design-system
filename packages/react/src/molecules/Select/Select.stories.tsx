import React from "react";

import Select from "./Select";
import "@ds.e/scss/lib/global.css";
import "@ds.e/scss/lib/Select.css";

const options = [
  {
    label: "Strict Black",
    value: "strict-black",
    key: "strict-black",
  },
  {
    label: "Heavenly Green",
    value: "heavenly-green",
    key: "heavenly-green",
  },
  {
    label: "Sweet Pink",
    value: "sweet-pink",
    key: "sweet-pink",
  },
];

export default {
  title: "Select",
};

export const Default = () => <Select options={options} />;
