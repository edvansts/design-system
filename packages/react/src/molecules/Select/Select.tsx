import React, {
  createRef,
  ReactNode,
  useEffect,
  useMemo,
  useRef,
  useState,
} from "react";

import Text from "../../atoms/Text";

const KEY_CODES = {
  ENTER: "Enter",
  SPACE: " ",
  ARROW_DOWN: "ArrowDown",
  ARROW_UP: "ArrowUp",
  ESC: "Escape",
};

function getNextOptionIndex(
  currentIndex: number | null,
  options: SelectOption[],
  key: string
) {
  if (currentIndex == null) {
    return 0;
  }

  if (key === KEY_CODES.ARROW_DOWN) {
    if (currentIndex == options.length - 1) {
      return 0;
    }

    return currentIndex + 1;
  }

  if (key === KEY_CODES.ARROW_UP) {
    if (currentIndex === 0) {
      return options.length - 1;
    }

    return currentIndex - 1;
  }

  return 0;
}

export interface SelectOption {
  label: string;
  value: string;
  key: string;
}

export interface RenderOptionProps {
  isSelected: boolean;
  option: SelectOption;
  getOptionRecommendedProps: (overrideProps?: Object) => Object;
}

export interface SelectProps {
  onOptionSelected?: (option: SelectOption, optionIndex: number) => void;
  options?: SelectOption[];
  placeholder?: string;
  renderOption?: (props: RenderOptionProps) => ReactNode;
}

function Select({
  options = [],
  placeholder = "Please select a option...",
  onOptionSelected: handler,
  renderOption,
}: SelectProps) {
  const [isOpen, setIsOpen] = useState(false);
  const [selectedOption, setSelectedOption] = useState<SelectOption | null>(
    null
  );
  const [hightlightedIndex, setHightlightedIndex] = useState<number | null>(
    null
  );

  const buttonRef = useRef<HTMLButtonElement>(null);
  const [optionRefs, setOptionRefs] = useState<React.RefObject<HTMLElement>[]>(
    []
  );

  const overlayTop = useMemo(() => {
    return (buttonRef.current?.offsetHeight ?? 0) + 10;
  }, [buttonRef.current?.offsetHeight]);

  const placeholderText = selectedOption?.label ?? placeholder;

  useEffect(() => {
    setOptionRefs(options.map(() => createRef<HTMLElement>()));
  }, [options.length]);

  useEffect(() => {
    if (hightlightedIndex !== null && isOpen) {
      const ref = optionRefs[hightlightedIndex];

      if (ref && ref.current) {
        ref.current.focus();
      }
    }
  }, [isOpen, hightlightedIndex]);

  function onOptionSelected(option: SelectOption, index: number) {
    if (handler) {
      handler(option, index);
    }

    setSelectedOption(option);
    setIsOpen(false);
  }

  function toggleSelectOpened() {
    setIsOpen((value) => !value);
  }

  function onButtonKeyDown(event: React.KeyboardEvent<HTMLButtonElement>) {
    event.preventDefault();

    if (
      [KEY_CODES.ARROW_DOWN, KEY_CODES.ENTER, KEY_CODES.SPACE].includes(
        event.key
      )
    ) {
      setIsOpen(true);

      hightlightOption(0);
    }
  }

  function onOptionKeyDown(event: React.KeyboardEvent<HTMLElement>) {
    if (event.key == KEY_CODES.ESC) {
      setIsOpen(false);
      return;
    }

    if ([KEY_CODES.ARROW_DOWN, KEY_CODES.ARROW_UP].includes(event.key)) {
      hightlightOption(
        getNextOptionIndex(hightlightedIndex, options, event.key)
      );
    }

    if (KEY_CODES.ENTER == event.key && hightlightedIndex !== null) {
      onOptionSelected(options[hightlightedIndex], hightlightedIndex);
    }
  }

  function hightlightOption(optionIndex: number | null) {
    setHightlightedIndex(optionIndex);
  }

  return (
    <div className="dse-select">
      <button
        data-testid="DseSelectButton"
        onKeyDown={onButtonKeyDown}
        className="dse-select__label"
        onClick={toggleSelectOpened}
        ref={buttonRef}
        aria-haspopup={true}
        aria-expanded={isOpen ?? undefined}
        aria-controls="dse-select-list"
      >
        <Text size="lg">{placeholderText}</Text>

        <svg
          xmlns="http://www.w3.org/2000/svg"
          className={`dse-select__caret ${
            isOpen ? "dse-select__caret--open" : "dse-select__caret--closed"
          }`}
          fill="none"
          viewBox="0 0 24 24"
          stroke="currentColor"
          width="1rem"
          height="1rem"
        >
          <path
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            d="M19 9l-7 7-7-7"
          />
        </svg>
      </button>

      {isOpen ? (
        <ul
          role="menu"
          id="dse-select-list"
          className="dse-select__overlay"
          style={{ top: overlayTop }}
        >
          {options.map((option, index) => {
            const isSelected = selectedOption?.key == option.key;
            const isHightlighted = index === hightlightedIndex;

            const ref = optionRefs[index];

            const className = `dse-select__option ${
              isSelected ? "dse-select__option--select" : ""
            }
            ${isHightlighted ? "dse-select__option--hightlighted" : ""}
            }`;
            const onClick = () => onOptionSelected(option, index);
            const key = option.key;

            const renderOptionsProps: RenderOptionProps = {
              option,
              isSelected,
              getOptionRecommendedProps: (overrideProps = {}) => ({
                role: "menuitemradio",
                "aria-label": option.label,
                "aria-checked": isSelected ? true : undefined,
                tabIndex: isHightlighted ? -1 : 0,
                onKeyDown: onOptionKeyDown,
                onMouseEnter: () => hightlightOption(index),
                onMouseLeave: () => hightlightOption(null),
                ref,
                className,
                onClick,
                key,
                ...overrideProps,
              }),
            };

            if (renderOption) {
              return renderOption(renderOptionsProps);
            }

            return (
              <li {...renderOptionsProps.getOptionRecommendedProps()}>
                <Text>{option.label}</Text>

                {isSelected ? (
                  <svg
                    xmlns="http://www.w3.org/2000/svg"
                    className="h-6 w-6"
                    fill="none"
                    viewBox="0 0 24 24"
                    stroke="currentColor"
                    width="1rem"
                    height="1rem"
                  >
                    <path
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      strokeWidth={2}
                      d="M5 13l4 4L19 7"
                    />
                  </svg>
                ) : null}
              </li>
            );
          })}
        </ul>
      ) : null}
    </div>
  );
}

export default Select;
