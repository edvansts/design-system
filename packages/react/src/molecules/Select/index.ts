export {
  default,
  SelectProps,
  SelectOption,
  RenderOptionProps,
} from "./Select";
