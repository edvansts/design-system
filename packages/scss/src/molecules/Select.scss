@import 'foundation/all';

.dse-select {
    display: flex;
    flex-direction: column;

    position: relative;

    &__caret {
        transition: all 250ms ease;

        &--open {
            transform: rotate(180deg);
        }

        &--closed {
            transform: rotate(0deg);
        }
    }

    &__label {
        display: flex;
        align-items: center;
        justify-content: space-between;

        @include padding('xs');
        border: 1px solid $form-border-color;
        background-color: $form-bg-color;
        font-family: $body-font-family;

        cursor: pointer;
    }

    &__overlay {
        @include padding('none');
        @include margin('none');
        @include shadow('base');

        list-style-type: none;

        border: 1px solid $form-border-color;
        background-color: $form-bg-color;
        z-index: 99;

        position: absolute;
        width: 100%;
        
    }


    &__option {
        display: flex;
        justify-content: space-between;
        align-items: center;
        width: 100%;
        cursor: pointer;
        box-sizing: border-box;

        @include padding('xs');

        &:hover {
            background-color: $form-bg-color-hover;
        }

        &--select {
            background-color: $form-bg-color-active;
            color: $form-color-active;

        }
    }
}

