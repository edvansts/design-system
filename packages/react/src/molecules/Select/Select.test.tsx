import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Select from "./Select";

const options = [
  {
    label: "Strict Black",
    value: "strict-black",
    key: "strict-black",
  },
  {
    label: "Heavenly Green",
    value: "heavenly-green",
    key: "heavenly-green",
  },
  {
    label: "Sweet Pink",
    value: "sweet-pink",
    key: "sweet-pink",
  },
];

describe("Select component", () => {
  it("renders all options passed to it", () => {
    render(<Select options={options} />);

    userEvent.click(screen.getByTestId("DseSelectButton"));

    expect(screen.getAllByRole("menuitemradio")).toHaveLength(options.length);
  });

  it("renders options using custom renderOption method if passed as prop", () => {
    render(
      <Select
        options={options}
        renderOption={({ option, getOptionRecommendedProps }) => {
          return (
            <li
              data-testid="CustomRenderOption"
              {...getOptionRecommendedProps()}
            >
              {option.label}
            </li>
          );
        }}
      />
    );

    userEvent.click(screen.getByTestId("DseSelectButton"));

    expect(screen.getAllByRole("menuitemradio")).toHaveLength(options.length);
  });

  it("calls the onOptionSelected prop with the selected option and its index if passed", () => {
    const onOptionSelected = jest.fn();

    render(
      <Select
        options={options}
        onOptionSelected={onOptionSelected}
        renderOption={({ option, getOptionRecommendedProps }) => {
          return (
            <li data-testid={option.key} {...getOptionRecommendedProps()}>
              {option.label}
            </li>
          );
        }}
      />
    );

    userEvent.click(screen.getByTestId("DseSelectButton"));
    userEvent.click(screen.getByTestId(options[1].key));

    expect(onOptionSelected).toBeCalledWith(options[1], 1);
  });

  it("the button label changes to the selected option label", () => {
    render(<Select options={options} />);

    const button = screen.getByTestId("DseSelectButton");

    userEvent.click(button);
    userEvent.click(screen.getAllByRole("menuitemradio")[1]);

    expect(button).toHaveTextContent(options[1].label);
  });

  it("snapshot of the selected option state", () => {
    const { asFragment } = render(<Select options={options} />);

    const button = screen.getByTestId("DseSelectButton");

    userEvent.click(button);
    userEvent.click(screen.getAllByRole("menuitemradio")[1]);

    expect(asFragment()).toMatchSnapshot();
  });

  it("snapshot of the base state", () => {
    const { asFragment } = render(<Select options={options} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("snapshot of the options menu open state", () => {});

  it("can customize select label", () => {});
});

test("snapshot of the options menu open state", () => {
  const { asFragment } = render(<Select options={options} />);

  userEvent.click(screen.getByTestId("DseSelectButton"));

  expect(asFragment()).toMatchSnapshot();
});

test("can customize select label", () => {
  const customLabel = "THIS IS A CUSTOM LABEL";
  render(<Select options={options} placeholder={customLabel} />);

  expect(screen.getByText(/THIS IS A CUSTOM LABEL/)).toBeInTheDocument();
});
