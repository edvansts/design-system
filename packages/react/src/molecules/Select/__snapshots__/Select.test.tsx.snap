// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component snapshot of the base state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="dse-select-list"
      aria-expanded="false"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
    >
      <p
        class="dse-text dse-font-size-md"
      >
        Please select a option...
      </p>
      <svg
        class="dse-select__caret dse-select__caret--closed"
        fill="none"
        height="1rem"
        stroke="currentColor"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Select component snapshot of the selected option state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="dse-select-list"
      aria-expanded="false"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
    >
      <p
        class="dse-text dse-font-size-md"
      >
        Heavenly Green
      </p>
      <svg
        class="dse-select__caret dse-select__caret--closed"
        fill="none"
        height="1rem"
        stroke="currentColor"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`snapshot of the options menu open state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="dse-select-list"
      aria-expanded="true"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
    >
      <p
        class="dse-text dse-font-size-md"
      >
        Please select a option...
      </p>
      <svg
        class="dse-select__caret dse-select__caret--open"
        fill="none"
        height="1rem"
        stroke="currentColor"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
    <ul
      class="dse-select__overlay"
      id="dse-select-list"
      role="menu"
      style="top: 10px;"
    >
      <li
        aria-label="Strict Black"
        class="dse-select__option 
            
            }"
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-font-size-md"
        >
          Strict Black
        </p>
      </li>
      <li
        aria-label="Heavenly Green"
        class="dse-select__option 
            
            }"
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-font-size-md"
        >
          Heavenly Green
        </p>
      </li>
      <li
        aria-label="Sweet Pink"
        class="dse-select__option 
            
            }"
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-font-size-md"
        >
          Sweet Pink
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
